src/main/kotlin/io/github/inductiveautomation/kindling/idb/tagconfig/model/ProviderStatistics.kt

```diff
@@ -0,0 +1,135 @@
+                }
+
+                valueSources.value.compute(node.statistics.dataSource ?: DEFAULT_VALUE_SOURCE) { _, value ->
+                    if (value == null) 1 else value + 1
+                }
+            }
+
+            node.statistics.isFolder -> totalFolders.value++
+            node.statistics.isUdtInstance -> totalUdtInstances.value++
+            node.statistics.isReadOnly == true -> totalReadOnlyTags.value++
+        }
+
+        if (node.statistics.historyEnabled == true) totalTagsWithHistory.value++
+
+        if (node.statistics.hasAlarms) {
+            totalTagsWithAlarms.value++
+            totalAlarms.value += node.statistics.numAlarms
+        }
+
+        if (node.statistics.hasScripts) {
+            totalTagsWithEnabledScripts.value++
+            totalEnabledScripts.value += node.statistics.numScripts
+        }
+    }
+
+    override fun toString(): String = buildString {
+        appendLine("Atomic Tags: $totalAtomicTags")
+        appendLine("- By Data Source")
+        appendLine(valueSources.value.entries.joinToString("\n|- ", prefix = "|- ") { (key, value) -> "$key: $value" })
+        appendLine("- By Data Type")
+        appendLine(dataTypes.value.entries.joinToString("\n|- ", prefix = "|- ") { (key, value) -> "$key: $value" })
+        appendLine("Folders: $totalFolders")
+        appendLine("Udt Instances: $totalUdtInstances")
+        appendLine("Udt Definition: $totalUdtDefinitions")
+        appendLine("Tags with History Enabled: $totalTagsWithHistory")
+        appendLine("Scripts:")
+        appendLine("- $totalEnabledScripts scripts found on $totalTagsWithEnabledScripts tags")
+        appendLine("Alarms:")
+        appendLine("- $totalAlarms alarms found on $totalTagsWithAlarms tags")
+        appendLine("Orphaned Tags: $totalOrphanedTags")
+    }
+
+    companion object {
+        private const val DEFAULT_DATA_TYPE = "Int4"
+        private const val DEFAULT_VALUE_SOURCE = "memory"
+    }
+
+    sealed class ProviderStatistic<T>(val name: String) {
+        abstract val value: T
+        val humanReadableName = name.splitCamelCase()
+            .replaceFirstChar { if (it.isLowerCase()) it.titlecase(Locale.getDefault()) else it.toString() }
+
+        operator fun getValue(thisRef: Any?, property: KProperty<*>): T = value
+
+        override fun toString(): String = value.toString()
+    }
+
+    class QuantitativeStatistic internal constructor(
+        name: String,
+        override var value: Int = 0,
+    ) : ProviderStatistic<Int>(name)
+
+    class MappedStatistic internal constructor(
+        name: String,
+        override val value: MutableMap<String, Int> = mutableMapOf(),
+    ) : ProviderStatistic<MutableMap<String, Int>>(name)
+
+    class ListStatistic<T> internal constructor(
+        name: String,
+        override val value: MutableList<T> = mutableListOf(),
+    ) : ProviderStatistic<MutableList<T>>(name)
+
+    class DependentStatistic<T, P> internal constructor(
```

Comment:
I like this, but wonder if there's some other way to approach this whole set of statistic dependencies (there may not be! I just haven't fully grokked everything)